### ISSUE TITLE
set anchor z center

### ISSUE THREAD

**Issue by zhashl** [traveller59/second.pytorch#274]
Hello Yan @traveller59 
      I'm using your code to train my own dataset. My dataset is in ground coordinate system. The coordinate of z axis is close to 0. I want to know the meaning of z center of each anchor so that I can set them by myself.
      And if I dense the 3D voxel feature when the number of z is 4 instead of 2, how should I set my z center of anchors?
     Thank you.

**Comment by micjabbour** [comment]
Sorry, I am not sure I can understand your question. What kind of anchor generator are you using? (anchor_generator_stride or anchor_generator_range)?

It might be more helpful if you can mention the exact name of the configuration parameter you are trying to refer to. . .

**Comment by zhashl** [comment]
> Sorry, I am not sure I can understand your question. What kind of anchor generator are you using? (anchor_generator_stride or anchor_generator_range)?
> 
> It might be more helpful if you can mention the exact name of the configuration parameter you are trying to refer to (along with the configuration file you are trying to modify). . .

It's my mistake that I didn't describe it exactly. The configuration parameter is “anchor_ranges” in configuration file. It's written as followed:
 target_assigner: {
      class_settings: {
        anchor_generator_range: {
          sizes: [1.6, 3.9, 1.56] # wlh
          # anchor_ranges: [0, -40.0, -1.00, 70.4, 40.0, -1.00] # carefully set z center
          **anchor_ranges:** [0, -32.0, **-1.00**, 52.8, 32.0, **-1.00**] # carefully set z center
          rotations: [0, 1.57] # DON'T modify this unless you are very familiar with my code.
        }
I wonder how to set the third and the last parameters (-1.00) in anchor_ranges.  Did you calculate the parameter or setting it by its actual physical meaning?
    In my dataset, I transform the points and label into "ground coordinate system" whose coordinate in z axis is close to 0. Can you please teach me how to set my "anchor_ranges"?

 Thank you ~~

**Comment by micjabbour** [comment]
I would say that anchors need to have a z coordinate that is as close as possible to objects of the anchor generator's class in your dataset. I think that you can obtain an acceptable value for this by calculating the average of the z coordinate of all objects from the corresponding class in your dataset. Since the plane z=0 corresponds to the ground in your dataset, the value can be set to the average height of objects of the corresponding class divided by 2 (assuming that all bounding boxes of that class touch the ground plane from below).

Hope this helps.

**Comment by pyaf** [comment]
According to the paper:
>Because the objects to be detected are of approximately fixed sizes, we use fixed-size anchors
determined based on the means of the sizes and center locations of all ground truths in the KITTI
training set with rotations of 0 and 90 degrees. For cars, we use an anchor with dimensions of
w = 1.6 × l = 3.9 × h = 1.56 m, centered at z = − 1.0 m. For pedestrians, we use an anchor with
dimensions of w = 0.6 × l = 0.8 × h = 1.73 m, and for cyclists, the anchor has dimensions of
w = 0.6 × l = 1.76 × h = 1.73 m; both are centered at z = − 0.6 m.
Each anchor is assigned a one-hot vector of classification targets, a 7-vector of box regression

for mean height of 1.56, they used z=-1.0, which contradicts what you say @micjabbour 
I'm unable to decide z-centers for my `anchor_ranges` :(

**Comment by johleh** [comment]
As long as you use the default BEV similarity to decide whether anchors are positive, negative or "don't cares" the z-center is only used for box regression. 

in this case, the choice of the z-center is not of great importance. the network will learn to correct it with a bias term in any case. -1 is simply the vertical center of the input space, which is restricted between -3 and 1.

**Comment by micjabbour** [comment]
@pyaf, I am sorry, I might have misunderstood. I don't see how does this exactly contradict my previous comment. The mean of the z coordinates of the centers of all objects of a specific class has nothing to do with their average height. Are you trying to correlate them with each other somehow?

**Comment by pyaf** [comment]
@johleh thanks, I understand the model is capable of adjusting the z-coordinate and z-center is not that important here. But my question is: in `anchor_ranges` we are required to provide two extremes of x, y, and z, right? for example, take this one:
`          anchor_ranges: [0, -32.0, -1.00, 52.8, 32.0, -1.00] # carefully set z center`
the extremums of x and y make sense, they got a span of 52.8 and 64 respectively. But the span of z is zero as the range of z is [-1, -1], which is a little bit confusing.

@micjabbour I'm sorry I should have clarified,  You say:
>I think that you can obtain an acceptable value for this by calculating the average of the z coordinate of all objects from the corresponding class in your dataset.

and

> the value can be set to the average height of objects of the corresponding class divided by 2 (assuming that all bounding boxes of that class touch the ground plane from below).

here my "value" you mean z-center, right? Your statements make sense, but in kitti dataset the average height of all cars is 1.56 so the z-center should be around 0.8, right? but in the paper, they took it as -1, which is a bit confusing.

I'm completely new to lidar-based 3d object detection, so, might have misunderstood something.

**Comment by micjabbour** [comment]
Oh, I think I understand your problem. The  KITTI dataset follows a coordinate system where the origin is located inside the LIDAR sensor (above the ground by about 1.73m), see [here](http://www.cvlibs.net/datasets/kitti/setup.php). My comment that you are citing above was in response to the OP's claim that they are following a "ground coordinate system" where the ground corresponds to the plane z=0.

Hope this helps...

**Comment by pyaf** [comment]
@micjabbour spot on! thank you so much!
One more question: suppose I've a new object category, say a double decker bus,with mean height 3 meters. now if I've to select z-center for this object category I'll go with ((3/2) - 1.73) = -0.23, right? 

Another question, what should be the `z-min` and `z-max` in `anchor_ranges`?

**Comment by micjabbour** [comment]
@pyaf, glad to help. This looks like a reasonable choice (assuming you are following a similar coordinate system to the one adopted by the KITTI dataset). To avoid coordinate system problems, you can also calculate the mean of the z-coordinate of all objects of your class of interest and use the result.

`z-min` and `z-max` should be set to the same value (the one we just discussed), since we don't want anchors to be stacked on top of each other across the z-axis (because all interesting objects are on the ground).

**Comment by pyaf** [comment]
Got it, thank you @micjabbour

**Comment by simaker** [comment]
I have the same Issue. I created my own labels for another dataset. Can i assume, that if the visualized gt-bounding-boxes in the kitti-viewer are correctly, that the anchors will generated correctly? Especially the coordinate transformations, box center, and so on

**Comment by micjabbour** [comment]
@simaker Apologizes for the late response, I just saw your comment.

No, visualizing annotated ground-truth data in the viewer has nothing to do with anchor generation and network training/prediction. You can verify the anchor generation only by dumping some of them and visualizing yourself to see if they line up with the point cloud scene you have (but this requires that you modify the relevant places of the code manually AFAIK). As previously mentioned in this thread, The z-center value is not of great importance, because the network is most likely able to learn this fixed correction value. If you are not sure about the co-ordinate system in your dataset, There are many ways to figure that out. E.g. you can just try to calculate the mean of the z-coordinate value of the centers of all the objects in the dataset, or maybe visualize the scene in a viewer that shows the origin and axis, ...

**Comment by hz3014** [comment]
Hi, @micjabbour 
I am trying to train my own custom dataset too. In order to check my generated groundtruth is correct, I checked the gt_database folder generated by create_data.py. The reason I am doing that is that many files are empty even though the boudning box is showing perfectly in my visualizer. 

I am quite confused what is stored in each file, they seem not to be point coordinates which is in the groundtruth bounding box because all the values are quite small(significant different from coordinate values). My guess was the distance to the centre of bounding box(in terms of xyz), but if so why dont they store the centre coordinate.

Any hint would be helpful!

**Comment by micjabbour** [comment]
Hi @hz3014, I am not sure I could understand what you are expecting to find in `gt_database`, but I'll try to explain its role in the hope that this clears things up.

The `gt_database` folder does not contain all the ground truth data used to train the neural network. It is just a database of all objects in your dataset. In other words, files in the mentioned directory store points that lie inside your ground truth annotations, translated so that the center of their box matches the origin (I think that this makes your guess in the end correct). Take a look at how they are generated [here](https://github.com/traveller59/second.pytorch/blob/3aba19c9688274f75ebb5e576f65cfe54773c021/second/data/all_dataset.py#L60-L68). One can visualize the points in these files using something like (make sure to have `pptk` installed and imported):

```python
>>> pc = np.fromfile('/path/to/gt_database/xxx_class_n.bin', dtype=np.float32).reshape(-1, 4)
>>> pc = pc[:, 0:3]  # drop reflectance from original points array
>>> color = np.ones(pc.shape, dtype=np.float32)  # color all points white
>>> pptk.viewer(pc[:, 0:3], color)  # display
```

This database is used to augment frames from the dataset while training, quoting the published paper "SECOND: Sparsely Embedded Convolutional Detection":

> The major problem we encountered during training was the existence of too few ground truths, which significantly limited the convergence speed and final performance of the network. To solve this problem, we introduced a data augmentation approach. First, we generated a database containing the labels of all ground truths and their associated point cloud data (points inside the 3D bounding boxes of the ground truths) from the training dataset. Then, during training, we randomly selected several ground truths from this database and introduced them into the current training point cloud via concatenation.

The relevant code that inserts these samples from the database into point cloud can be found in the file `second/core/sample_ops.py`.

Hope this helps.
